### PR TITLE
Make CSI volume attributes first class

### PR DESCRIPTION
**Pull request by musse**
**What this PR does / why we need it**:
Move CSI volume attributes from PV annotation to CSI volume source first class field

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by saad-ali** [comment]
This looks like the correct approach to me. Need to update the auto-generated API code. Once this is merged, will need to make the corresponding changes to the external components (external-attacher, etc.).

/assign @vladimirvivien, @chakri-nelluri

File: pkg/volume/csi/csi_mounter.go [hunk 1]
```diff
@@ -302,20 +302,13 @@ func getVolAttribsFromSpec(spec *volume.Spec) (map[string]string, error) {
...
-	}
-	jsonAttribs := annotations[csiVolAttribsAnnotationKey]
-	if jsonAttribs == "" {
-		return nil, nil // csi annotation not found
-	}
-	attribs := map[string]string{}
-	if err := json.Unmarshal([]byte(jsonAttribs), &attribs); err != nil {
-		glog.Error(log("error parsing csi PV.Annotation [%s]=%s: %v", csiVolAttribsAnnotationKey, jsonAttribs, err))
-		return nil, err
+
```
Comment (vladimirvivien) [inline pkg/volume/csi/csi_mounter.go]: Function `getVolAttribsFromSpec()` found [here]( https://github.com/musse/kubernetes/blob/566ffc53a9b8bf9860b3b76f366f8150b90e3b4c/pkg/volume/csi/csi_mounter.go#L157) . Variable `csiSource`, declared [here](https://github.com/musse/kubernetes/blob/566ffc53a9b8bf9860b3b76f366f8150b90e3b4c/pkg/volume/csi/csi_mounter.go#L113), should already have the attributes (which is what the function is doing again).

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -1736,6 +1736,10 @@ type CSIPersistentVolumeSource struct {
 	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
 	// +optional
 	FSType string `json:"fsType,omitempty" protobuf:"bytes,4,opt,name=fsType"`
+
+	// Attributes of the volume to publish.
+	// +optional
+	VolumeAttributes map[string]string `json:"volumeAttributes,omitempty" protobuf:"bytes,5,rep,name=volumeAttributes"`
```
Comment (vladimirvivien) [inline staging/src/k8s.io/api/core/v1/types.go]: As Saad mentioned, don't forget to re-generate the API code for this type.  If you need help, let me know.

Comment (vladimirvivien) [reply staging/src/k8s.io/api/core/v1/types.go]: @musse When you change the API object, you must re-generate API source code.  Here are some steps:

0) Make sure to make you API change in `pkg/apis/core/v1/types.go` also.
1) From the kubernetes root dir, run `make update`
2) Follow directions when that task fails, it should walk you through what needs to be done
2) Run `make verify` to ensure all is good.  Follow directions there when things are broken.

Post here (or ping me on slack) if things dont work.

Comment (musse) [reply staging/src/k8s.io/api/core/v1/types.go]: I don't have the `pkg/apis/core/v1/types.go` file. Will it be generate with `make update`?

File: pkg/volume/csi/csi_mounter.go (hunk 1 above)
Comment (musse) [reply pkg/volume/csi/csi_mounter.go]: Done, let me know if the change is not what you expected.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 2 above)
Comment (vladimirvivien) [reply staging/src/k8s.io/api/core/v1/types.go]: @musse Apologies, this is the file and line# https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1607

File: pkg/volume/csi/csi_plugin.go [hunk 3]
```diff
@@ -31,8 +31,7 @@ import (
 )
 
 const (
-	csiPluginName              = "kubernetes.io/csi"
-	csiVolAttribsAnnotationKey = "csi.volume.kubernetes.io/volume-attributes"
+	csiPluginName = "kubernetes.io/csi"
```
Comment (vladimirvivien) [inline pkg/volume/csi/csi_plugin.go]: @musse this should already be there, why adding it here?

**Review by vladimirvivien — COMMENTED** [review]
@musse this looks good. I did leave one comment.

File: pkg/volume/csi/csi_plugin.go (hunk 3 above)
Comment (musse) [reply pkg/volume/csi/csi_plugin.go]: I only removed the spaces before the `=` sign.

**Comment by saad-ali** [comment]
/lgtm
/approve

For API approval:
/assign @thockin

**Comment by saad-ali** [comment]
@musse could you rebase this PR.

**Comment by musse** [comment]
@saad-ali Done.

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -1634,6 +1634,10 @@ type CSIPersistentVolumeSource struct {
 	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
 	// +optional
 	FSType string
+
+	// Attributes of the volume to publish.
+	// +optional
+	VolumeAttributes map[string]string
```
Comment (thockin) [inline pkg/apis/core/types.go]: why does this want the "Volume" prefix?  Does it feel redundant in actual use?

Comment (saad-ali) [reply pkg/apis/core/types.go]: It's to align with the parameter `volume_attributes` in the CSI spec.

**Comment by saad-ali** [comment]
/lgtm
Re-adding LGTM after rebase

**Pull request merged**